File: src/main/java/com/utopia/Region.java
package com.utopia;

/**
 * A class that defines a rectangular region in a larger space.
 */
public class Region {
    public int startX = 0;
    public int startY = 0;
    public int width = 0;
    public int height = 0;

    public Region() {}
    
    public Region(int startX, int startY, int width, int height) {
        this.startX = startX;
        this.startY = startY;
        this.width = width;
        this.height = height;
    }

    public int getEndX() {
        return startX + width;
    }

    public int getEndY() {
        return startY + height;
    }

    public boolean contains(int x, int y) {
        return (x >= startX &&
                y >= startY &&
                x <= getEndX() &&
                y <= getEndY());
    }

    public boolean contains(final Region other) {
        return (other.startX >= startX &&
                other.startY >= startY &&
                other.getEndX() <= getEndX() &&
                other.getEndY() <= getEndY());
    }

    @Override public boolean equals(Object other) {
        if (this == other) {
            return true;
        }

        if (other == null) {
            return false;
        }

        if (getClass() != other.getClass()) {
            return false;
        }

        final Region otherRegion = (Region)other;
        return (this.startX == otherRegion.startX &&
                this.startY == otherRegion.startY &&
                this.width == otherRegion.width &&
                this.height == otherRegion.height);
    }
    
    @Override public String toString() {
        return "(" + startX + "," + startY + "," + width + "," + height + ")";
    }
}

File: frontend/src/region.js
////////////////////////////////////////////////////////////////////////////////

/**
 * A class that defines a rectangular region in a larger space.
 */
export class Region {
    constructor(startX = 0, startY = 0, width = 0, height = 0) {
        this.startX = startX;
        this.startY = startY;
        this.width = width;
        this.height = height;
    }

    get endX() {
        return this.startX + this.width;
    }

    get endY() {
        return this.startY + this.height;
    }

    // MYTODO minimum here and in java shouldnt' be inclusive? THIS but
    // not region. update java.
    contains(x, y) {
        return (x >= this.startX &&
                y >= this.startY &&
                x < this.endX &&
                y < this.endY);
    }

    containsRegion(other) {
        return (other.startX >= this.startX &&
                other.startY >= this.startY &&
                other.endX <= this.endX &&
                other.endY <= this.endY);
    }
}
